pkg/kubelet/kubelet_node_status.go

```diff
@@ -227,6 +229,21 @@ func (kl *Kubelet) initialNode() (*v1.Node, error) {
 		}
 		nodeTaints = append(nodeTaints, taints...)
 	}
+
+	unschedulableTaint := v1.Taint{
+		Key:    algorithm.TaintNodeUnschedulable,
+		Effect: v1.TaintEffectNoSchedule,
+	}
+
+	// If TaintNodesByCondition enabled, taint node with TaintNodeUnschedulable when initializing
+	// node to avoid race condition; refer to #63897 for more detail.
+	if utilfeature.DefaultFeatureGate.Enabled(features.TaintNodesByCondition) {
+		if node.Spec.Unschedulable &&
+			!taintutil.TaintExists(nodeTaints, &unschedulableTaint) {
```

Comment:
If the taint exists with a different effect, this will duplicate it. Granted, that's probably a bug anyway, but I think it's safer to just check for the key here (overwriting as necessary). Although actually, does the NodeRestriction controller allow the taint effect to be changed here?
@liggitt

Reply:
initialNode() is only used to create the node in initial registration. it does not update nodes.

Reply:
> If the taint exists with a different effect, this will duplicate it. 

Yes, that up to user's setting; it's an valid case that user also add the effect, we only need to handle `NoSchedule` by kubelet.

> Although actually, does the NodeRestriction controller allow the taint effect to be changed here?

As liggitt@ said, `NodeRestriction controller` only reject node taints updates; this's accepted as creation :)

Reply:
If that's the case, then why check `TaintExists` at all here?

Reply:
They could've specified it explicitly in the initial taints flag

Reply:
Sorry, I'm confused... then couldn't this duplicate the taint specified from --register-with-taints?

Reply:
If it's duplicated, it'll not add it; that's why we check `TaintExists` :). And we only check against `unschedulableTaint`, which is handled `NodeLifeCycleController`; the other customized taint will follow the default Taint/Toleration behaviour :)

Reply:
To clarify:

1. Kubelet is started with `--register-with-taints=node.kubernetes.io/unschedulable=true:PreferNoSchedule`
2. This code checks for a taint with `node.kubernetes.io/unschedulable:NoSchedule`
3. No match is found, and the kubelet ends up with the duplicate taint `node.kubernetes.io/unschedulable:PreferNoSchedule` and `node.kubernetes.io/unschedulable:NoSchedule`

Reply:
@tallclair , `node.kubernetes.io/unschedulable:PreferNoSchedule` is used for priorities, and `node.kubernetes.io/unschedulable:NoSchedule` is used for predicates; so I think they're not duplicated; and `node.kubernetes.io/unschedulable:NoSchedule` will be removed when `.spec.unschedulable` is updated to `false`.